django/db/models/query.py

```diff
@@ -1137,6 +1130,21 @@ def _has_filters(self):
         """
         return self.query.has_filters()
 
+    def _check_allowed(self, action):
+        require_no_limits = {
```

Comment:
The dict can be removed if this method is only called by those methods which have to make sure slice has not been taken or union, intersection, difference has not been applied.
Then this method can be shorten to.

```
def check_queryset_method_allowed(self, action_to_msg, slice_check=True, combinator_check=True):
        if slice_check:
            assert not self.query.has_limit(), 
                'Cannot use %s ' % action_to_msg + 'once a slice has been taken'

        if combinator_check:
            assert not self.query.combinator==None , 
                'Cannot use %s ' % action_to_msg + 'once %s has been applied' % self.query.combinator
```

Now methods like _filter_or_exclude, count, extra etc will have to just call this method to make sure they can execute.
As we only have restrictions on query combinators (union, intersection and difference) and slice, the two keyword arguments are passed to this method.